### PR TITLE
LoRaWAN Technology Section

### PR DESCRIPTION
**Pull request by hasnainvirk**
This is just for preview. Let's go through the material and diagrams first. @SeppoTakalo @AnttiKauppila @kivaisan @MarceloSalazar @janjongboom @AnotherButler

### CODE REVIEW

File: docs/reference/technology/connectivity/test-lora.md [hunk 1]
```diff
@@ -0,0 +1,291 @@
+# Mbed LoRaWAN Solution
+
+## What is LoRaWAN & Why ?
+The basic difference between "internet" and "internet of things (IoT)"  is that for IoT, there is just 'less of everything' available for a given network device: less memory, less processing capacity, less leg-room in terms of radio spectrum availability, less storage capacity and ofcourse less available power. These restrictions appear either because most of the "things" which will constitute IoT are battery powered and maximizing battery life-time is a priority, or they are supposed to be deployed in massive amounts, we are talking billions and trillions of tiny devices here.
```
[the remark this answers is not in the document]
Comment (hasnainvirk) [reply docs/reference/technology/connectivity/test-lora.md]: ofcourse is correct spelling I guess in British English :)

File: docs/reference/technology/connectivity/test-lora.md [hunk 2]
```diff
@@ -0,0 +1,291 @@
...
+
+Usually the network topolgy looks like a one-hop star netwotk, however there may be cases when there is a repeater involved in the radio path working as a middle-man between Base Station and the Device. More than one repeater is not allowed by the current standard specification.  
+
+## LoRaWAN Standard Specifications
+
+LoRa Alliance is responsible for standardizing LoRaWAN protocol and until now there has been four officially released standard specifications.
+
+* LoRaWAN Specification v1.0
+* LoRaWAN Specification v1.0.1
+* LoRaWAN Specification v1.0.2
```
[the remark this answers is not in the document]
Comment (hasnainvirk) [reply docs/reference/technology/connectivity/test-lora.md]: It's not released publically yet. Anyhow, I think we can do away with mention of all subversions.

File: docs/reference/technology/connectivity/test-lora.md [hunk 3]
```diff
@@ -0,0 +1,291 @@
...
+ 
+##### LoRaWANStack class 
+
+`LoRaWANStack` class is the supervisory unit. It runs the state machine of the stack. At one end it provides a glue layer for the `LoRaWANInterface` which in turn provides a network interface to  the application, and at the other end it controls the division of labour among other units of the system. This class is responsible for handling interrupts generated by the radio driver, managing states and delegating jobs asked by the application to the next lower unit, i.e., `LoRaMac` class. 
+
+##### LoRaMac class
+`LoRaMac` class constitutes the core MAC functionality. It performs the operations delegated by the `LoRaWANStack` and hoists appropriate  flags for various MAC indications or confirmations which need further post processing. This class is also resposible for keeping track of timers using `LoRaWANTimer`, performing crypto operations using `LoRaMacCrypto`, processing MAC commands using `LoRaMacCommand` and processing channel plans using `LoRaMacChannelPlan` classes. In a nutshell, `LoRaMac` class processes jobs using smaller units, provides indications or confirmations of the operations performed to the `LoRaWANStack`.
```
[the remark this answers is not in the document]
Comment (hasnainvirk) [reply docs/reference/technology/connectivity/test-lora.md]: hoists is the correct word here.

**Comment by hasnainvirk** [comment]
@SeppoTakalo @AnotherButler Is it possible semantically to show this as a separate page ? Are there any engine related restrictions here ? Please verify if you agree with the next steps I propose here:

[I don't know why and when somebody removed any links to these sections from the text.]
1) Bringing back lorawanevents.md in the api/connectivity/lorawan section. 
2) Bringing back errorcodes.md
3) Bringing back lorawanconfig.md

Regarding lorawanconfig.md, I would like to point out that the blob in docs/configuration/connectivity doesn't impart any useful information as it is incoherent. I would like to revive all the sections mentioned above in the main text of lorawan.md.

**Comment by hasnainvirk** [comment]
@AnotherButler Could you please go ahead and move technology/connectivity/intro-to-lora.md to the tutorial section ? Then we can make this commit a proper pull request

File: docs/reference/technology/connectivity/test-lora.md [hunk 4]
```diff
@@ -0,0 +1,289 @@
...
+
+## LoRaWAN Network Architecture
+ 
+ A LoRaWAN network consists of three basic network elements:
+ 
+ 1. Device
+ 2. Base station 
+ 3. Network Server
+ 
+ Base station is essentially transparent, its job is simply to speak LoRa with the devices in its coverage area. The real network control lies in the cloud, i.e., the Network Server. However, there are many solutions available where Network Server also resides on the base station, making a non-scalable private LoRaWAN network.
```
Comment (janjongboom) [inline docs/reference/technology/connectivity/test-lora.md]: "there are many solutions available where Network Server also resides on the base station" -> put a big fat warning there that this is not LoRaWAN compliant.

File: docs/reference/technology/connectivity/test-lora.md [hunk 5]
```diff
@@ -0,0 +1,289 @@
...
+Each beacon is preceded by Beacon Guard time and no ping slot can be placed in that time period. A ping slot is a 30ms long unit which can be assigned to a Class B device. Beacon Reserved is a time period when the actual beacon is sent. Beacon Window is the time period when ping slots can be opened and assigned. 
+
+* Beacon Period = 128 seconds
+* Beacon Reserved = 2.120 seconds
+* Beacon Guard = 3 seconds
+* Beacon Window = 122.880 seconds
+
+In total there can be 4096, 30ms long ping slots in the Beacon Window. 
+
+**NOTE: Currently Mbed LoRaWAN Stack doesn't support Class B.**
```
Comment (janjongboom) [inline docs/reference/technology/connectivity/test-lora.md]: Make a note that Class B support is only feasible with LoRaWAN 1.1

File: docs/reference/technology/connectivity/test-lora.md [hunk 6]
```diff
@@ -0,0 +1,289 @@
...
+
+### Class C
+
+Class C devices are supposed to be main powered or that kind of devices who have sufficient amount of power supply available. These devices tend to remain in listening mode always when they are not transmitting.
```
Comment (janjongboom) [inline docs/reference/technology/connectivity/test-lora.md]: Or when you want to do multicast, f.e. firmware updates.

**Comment by janjongboom** [comment]
I feel this article is a bit unbalanced. The tech section at the end is good, but the beginning is a mix of basic IoT explanation mixed with deep LoRa information. I'd skip all generic IoT talk and rather explain a bit more about LoRaWAN (I can help if needed).

File: docs/reference/technology/connectivity/test-lora.md (hunk 4 above)
Comment (loverdeg-ep) [reply docs/reference/technology/connectivity/test-lora.md]: @janjongboom I figured what you are saying would eventually be specified. However I've never seen any specifications that define particulars about the network server or application server other then what is said in the LoRaWAN spec about encrypting mac and application data and providing app ID.

My last spec read through was  sometime in 2017 though. Could you point me to where this is specified?

Comment (loverdeg-ep) [reply docs/reference/technology/connectivity/test-lora.md]: @janjongboom My thinking behind this is that in order to do edgework with LoRa you'll need a network server on the local network. In this case I'd think gateways could forward traffic to both the local network server and a cloud network server for 'telemetry' purposes.

Base on the above I think there are some scenario's where having a LoRaWAN gateway and a LoRaWAN network server residing on the same hardware might be convenient.
The local network server would then forward application traffic to an application server elsewhere in the network.

Feel free to continue this conversation elsewhere if you'd prefer. My email should be in my github profile.

File: docs/reference/technology/connectivity/test-lora.md [hunk 7]
```diff
@@ -0,0 +1,289 @@
+# Mbed LoRaWAN Solution
+
+## What is LoRaWAN & Why ?
+The basic difference between "internet" and "internet of things (IoT)" is that for IoT, there is just 'less of everything' available for a given network device: less memory, less processing capacity, less leg-room in terms of radio spectrum availability, less storage capacity and ofcourse less available power. These restrictions appear either because most of the "things" which will constitute IoT are battery powered and maximizing battery life-time is a priority, or they are supposed to be deployed in massive amounts, we are talking billions and trillions of tiny devices here.
+
+This drive to “do more with less” leads to constraints that limit the applicability of traditional cellular networks, as well as of technologies, such as WiFi, due to energy and scalability requirements. A new class of wireless networking protocols have emerged to tackle the challanges posed by IoT which is generally known as Low Power Wide Area Networks (LPWANs).
```
[the remark this answers is not in the document]
Comment (hasnainvirk) [reply docs/reference/technology/connectivity/test-lora.md]: It's not copied. The article you referred to is a scientific publication which falls in public domain. I had read that article a few months ago. While writing this text, there is certainly an influence of those articles I read. Is it actually wrong to reuse the knowledge imparted by somebody else's work ?. Maybe a reference to the original article could be added here.

File: docs/reference/technology/connectivity/test-lora.md [hunk 8]
```diff
@@ -0,0 +1,289 @@
+# Mbed LoRaWAN Solution
+
+## What is LoRaWAN & Why ?
+The basic difference between "internet" and "internet of things (IoT)" is that for IoT, there is just 'less of everything' available for a given network device: less memory, less processing capacity, less leg-room in terms of radio spectrum availability, less storage capacity and ofcourse less available power. These restrictions appear either because most of the "things" which will constitute IoT are battery powered and maximizing battery life-time is a priority, or they are supposed to be deployed in massive amounts, we are talking billions and trillions of tiny devices here.
```
[the remark this answers is not in the document]
Comment (hasnainvirk) [reply docs/reference/technology/connectivity/test-lora.md]: IoT is the pretext to LoRaWAN. If this pretext was not there, existing internet technologies would have been enough to constitute an IoT as well.

File: docs/reference/technology/connectivity/test-lora.md (hunk 4 above)
Comment (hasnainvirk) [reply docs/reference/technology/connectivity/test-lora.md]: There is no concept of slave and master network servers in the spec. So if you have a network server residing on Base Station, that's the end of your network. It's a single star network. You can't interconnect various stars to make a big network. That's what I meant by being non-scalable.

File: docs/reference/technology/connectivity/test-lora.md [hunk 9]
```diff
@@ -0,0 +1,289 @@
...
+Usually the network topology looks like a one-hop star network, however there may be cases when there is a repeater involved in the radio path working as a middle-man between Base Station and the Device. More than one repeater is not allowed by the current standard specification.  
+
+## LoRaWAN Standard Specifications
+
+LoRa Alliance is responsible for standardizing LoRaWAN protocol and until now there has been two major officially released standard specifications and a few fix versions.
+
+* LoRaWAN Specification v1.0.X
+* LoRaWAN Specification v1.1
+
+Specification v1.0 and v1.0.X are essentially similar with a few differences. The latest specification v1.1 is largely different in terms of how network control expresses itself. The most important feature in v1.1 is the support for roaming. All the specifications are supposed to be backwards compatible. These standard documents are augmented with Regional Parameters Specifications which address the regional constraints pertaining to radio regulations across the world.
```
[the remark this answers is not in the document]
Comment (hasnainvirk) [reply docs/reference/technology/connectivity/test-lora.md]: That's actually true. I will amend this piece of text.

File: docs/reference/technology/connectivity/test-lora.md (hunk 5 above)
Comment (hasnainvirk) [reply docs/reference/technology/connectivity/test-lora.md]: Not really. 1.0.3 is specifically designed to entertain Class B devices. Rudiments of these devices is included in 1.0.2 as well.

File: docs/reference/technology/connectivity/test-lora.md (hunk 6 above)
Comment (hasnainvirk) [reply docs/reference/technology/connectivity/test-lora.md]: It's a general statement that covers both.

File: docs/reference/technology/connectivity/test-lora.md [hunk 10]
```diff
@@ -0,0 +1,289 @@
...
+
+<span class="images">![](https://s3-us-west-2.amazonaws.com/mbed-os-docs-images/class_c_timing.png)<span>Figure 4: Class C timing diagram</span></span>
+
+## LoRaWAN Connection Types
+There are two methods defined by the LoRaWAN specification for connecting to an access network.
+
+* ABP (Activation by personalization)
+* OTAA (Over the air activation) 
+
+### ABP (Activation by personalization)
```
[the remark this answers is not in the document]
Comment (hasnainvirk) [reply docs/reference/technology/connectivity/test-lora.md]: That's a bit presumptuous in my opinion. If the keys reside in a secure element than probably the most suitable of activation methods is ABP. You don't need to negotiate your way in. Yes, OTAA can give you the liberty, but still I think there will be a large number consumer devices provided by the service provider with factory burnt personalization keys.

File: docs/reference/technology/connectivity/test-lora.md [hunk 11]
```diff
@@ -0,0 +1,289 @@
...
+
+## LoRaWAN Data Message Types
+There are two types of data messages that can be sent using LoRaWAN protocol. 
+
+* Unconfirmed messages (fire and forget type of datagrams)
+* Confirmed messages (requires acknowledgement) from the other end.
+
+For Confirmed messages, going out to the network, i.e., transmitted by the device, an acknowledgement should be sent by the Network Server within the two RX windows opened by the Device. If no ack was received, the Device should wait for a time period equal to a given ack timeout and then retry.
+
+## Arm Mbed LoRaWAN Stack
```
[the remark this answers is not in the document]
Comment (hasnainvirk) [reply docs/reference/technology/connectivity/test-lora.md]: Mbed is the ecosystem name. The Stack is thus Mbed LoRaWAN Stack which happens to reside in Mbed OS.

**Comment by hasnainvirk** [comment]
@janjongboom Please go ahead and add what you think is suitable. I can review it once you have added stuff here. Please note that the essence I am trying to use here is Why, What and How.
I understand that you may have looked at these things millions of times before so, you just want to get to the point. I have written this assuming the reader is there to understand Why LoRaWAN, What LoRaWAN and How LoRaWAN. If you can keep the sanctity of this premise and still add useful stuff in the text, I will be utterly pleased.

File: docs/reference/technology/connectivity/test-lora.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (MarceloSalazar) [reply docs/reference/technology/connectivity/test-lora.md]: The first part is too generic and doesn't feel right. Let's focus on Lora and remove bluff from docs.

File: docs/reference/technology/connectivity/test-lora.md [hunk 12]
```diff
@@ -0,0 +1,289 @@
...
+
+Arm is committed to bring about the next revolution in technology by enabling IoT, Machine Learning and AI by taking up technological challanges and solving the problems of today as well as tomorrow.
```
[the remark this answers is not in the document]
Comment (MarceloSalazar) [reply docs/reference/technology/connectivity/test-lora.md]: I agree - no need to marketing stuff here.

**Comment by MarceloSalazar** [comment]
Thanks @hasnainvirk for the great work.

Just a note, the doc seems to have a good intro to the Lora technology.
However, there are many references to API and source code, which to me feels should go to the API or tutorials section. This is important to ensure we put the right things into the corresponding sections.
Have a look at other technologies: IP (Eth/Wifi/6lowpan)

@AnotherButler can you please help on this?

**Comment by hasnainvirk** [comment]
@MarceloSalazar The references are to the docs of the source code mainly. Here I am attempting to give overview of the stack itself and how it is tethered together.  Because we don't have a state machine diagram for Radio driver, I added some example code. Other entities are manifested by flow diagrams and links to their documentation.

**Comment by MarceloSalazar** [comment]
As per discussion with @AnotherButler , she's aware there might be a need move things around a bit in the future. There is an intention to have guidelines in place to help in understanding what goes where.
For now, this is considered good enough. Thanks!

**Comment by hasnainvirk** [comment]
@AnotherButler Yeah, that looks good. Please rename the .md file to reflect what it is. I named it test-lora.md as it was a test for me at the time. But if you are certainly moving intro-to-lora.md to Tutorial section, this file can have a different name.

**Pull request merged**